Implement pg_search

kljuni opened the pull request:
**1. Added pg_search Gem:**

- Integrated the pg_search gem to enhance search functionality within the application.

**2. Search Implementation:**

- Implemented tsearch and trigram (with two thresholds) search for improved search accuracy.

- Potential for further optimization by consolidating to a single threshold for trigram search in future iterations.

**3. Removed Algolia:**

- Deprecated the use of Algolia for search functionalities in favor of pg_search.

nmondollot reviewed (commented):
Wow, I'm very impressed!
Excellent job! 🤩 

It's almost shippable as is, but I still have a few nitpicks:
- we can make the code a bit simpler (and easier to maintain) if we stick more to the "Stimulus way" using targets and actions (cf my inline comments)
  - we could even go deeper and use a turbo frame with HTML instead of JSON (since we don't need a search api for the foreseeable future), but honestly I'm fine with json too
- the low threshold seems a bit too low for area search
  - eg. a query for a problem like `Duel` returns areas like `Potala`, `Restant du Long Rocher` and `Rocher du Général` => not very helpful
  - but I guess we can tweak the values later
- Can we disable autocorrect on the search textbox?

A bit more advanced:
- When the server is slow to answer, if I do 2 searches consecutively I don't know if the results being displayed are for my first search or the second => it's a bit confusing.
  - how does algolia handle this? do they "cancel" previous searches?
- It'd be nice to have better accessibility for desktop: up/down arrows, enter to select, escape to close the modal, etc

nmondollot commented:
Just had a look, it works really well 🤩 
Let's ship it when you're back!

Thank you for such great work 🙏

nmondollot commented on Gemfile:
```diff
@@ -15,7 +15,8 @@ gem 'sidekiq', '~> 7.1.2'
 gem 'bugsnag', '~> 6.25.2'
-gem 'algoliasearch-rails', '~> 2.2.2'
+gem 'pg_search', '~> 2.3.6'
+gem 'concurrent-ruby', '~> 1.2.3'
```
Why do you need the `concurrent-ruby` gem?

kljuni commented:
Hi Nicolas,

Thanks for the feedback! I am back from vacation. Today I added a small fix
for managing focus when the mouse is already over some item when we input a
new query. I also removed the redundant concurrent-ruby (I used it in the
past for a possible solution).

All the best,
Ivan

On Tue, 27 Aug 2024 at 13:51, Nicolas Mondollot ***@***.***>
wrote:

> Just had a look, it works really well 🤩
> Let's ship it when you're back!
>
> Thank you for such great work 🙏
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/boolder-org/boolder-rails/pull/47#issuecomment-2312351867>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AMIE2SONMN3ADORA4P225OTZTRR5XAVCNFSM6AAAAABJPZOXNWVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDGMJSGM2TCOBWG4>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>

nmondollot commented:
It's live on prod!
It seems to work pretty well 👍 
I'll make more tests in the upcoming days.

Thank you so much for your contribution ❤️
